### PR TITLE
Various cgroup and mounting fixes and optimizations

### LINKED ISSUE #3689: Kubernetes Information on tracee events doesn't appear on EKS
## Description
 
Hey,
I used installed tracee helm chart on eks version 1.27 and 1.25 
The helm charts deployed tracee version 0.18.1.
To do so I followed the instructions found here - https://aquasecurity.github.io/tracee/v0.19/tutorials/k8s-policies/
I added to tracee policy "sched_process_exec" event and restarted the daemonset. 
As you can see from the logs the Kubernetes information is empty for every event/signature.

I tried using global scope (the default) and tried changing it to container scope and nothing helped.

## Output of `tracee logs`:

```
{"level":"error","ts":1699967074.7665052,"msg":"error triggering container enrich in control plane","error":"containers.(*Containers).EnrichCgroupInfo: container already deleted"}

{"timestamp":1699967074786654029,"threadStartTime":1699967074754526405,"processorId":1,"processId":1,"cgroupId":1498,"threadId":1,"parentProcessId":741122,"hostProcessId":806569,"hostThreadId":806569,"hostParentProcessId":741122,"userId":0,"mountNamespace":4026532537,"pidNamespace":4026532538,"processName":"sh","executable":{"path":""},"hostName":"test-pod-1","containerId":"72ce9c01309449fcbe7374f7a3b76b9d561a5969b7ac8f5374405186858619a6","container":{"id":"72ce9c01309449fcbe7374f7a3b76b9d561a5969b7ac8f5374405186858619a6"},"kubernetes":{},"eventId":"713","eventName":"sched_process_exec","matchedPolicies":["signature-events"],"argsNum":11,"returnValue":0,"syscall":"execve","stackAddresses":null,"contextFlags":{"containerStarted":true,"isCompat":false},"threadEntityId":1076268326,"processEntityId":1076268326,"parentEntityId":2561041611,"args":[{"name":"cmdpath","type":"const char*","value":"/usr/bin/sh"},{"name":"pathname","type":"const char*","value":"/usr/bin/bash"},{"name":"dev","type":"dev_t","value":149},{"name":"inode","type":"unsigned long","value":6884317},{"name":"ctime","type":"unsigned long","value":1699952785706805776},{"name":"inode_mode","type":"umode_t","value":33261},{"name":"interpreter_pathname","type":"const char*","value":null},{"name":"interpreter_dev","type":"dev_t","value":null},{"name":"interpreter_inode","type":"unsigned long","value":null},{"name":"interpreter_ctime","type":"unsigned long","value":null},{"name":"argv","type":"const char**","value":["sh","-c","sleep 999"]},{"name":"interp","type":"const char*","value":"/usr/bin/sh"},{"name":"stdin_type","type":"string","value":"S_IFCHR"},{"name":"stdin_path","type":"char*","value":"/dev/null"},{"name":"invoked_from_kernel","type":"int","value":0},{"name":"env","type":"const char**","value":null}]}

{"timestamp":1699967074790669995,"threadStartTime":1699967074754526405,"processorId":1,"processId":1,"cgroupId":1498,"threadId":1,"parentProcessId":1,"hostProcessId":806569,"hostThreadId":806569,"hostParentProcessId":741122,"userId":0,"mountNamespace":4026532537,"pidNamespace":4026532538,"processName":"sleep","executable":{"path":""},"hostName":"test-pod-1","containerId":"72ce9c01309449fcbe7374f7a3b76b9d561a5969b7ac8f5374405186858619a6","container":{"id":"72ce9c01309449fcbe7374f7a3b76b9d561a5969b7ac8f5374405186858619a6"},"kubernetes":{},"eventId":"713","eventName":"sched_process_exec","matchedPolicies":["signature-events"],"argsNum":15,"returnValue":0,"syscall":"execve","stackAddresses":null,"contextFlags":{"containerStarted":true,"isCompat":false},"threadEntityId":1076268326,"processEntityId":1076268326,"parentEntityId":2561041611,"args":[{"name":"cmdpath","type":"const char*","value":"/usr/bin/sleep"},{"name":"pathname","type":"const char*","value":"/usr/bin/coreutils"},{"name":"dev","type":"dev_t","value":149},{"name":"inode","type":"unsigned long","value":6291895},{"name":"ctime","type":"unsigned long","value":1699952785730805852},{"name":"inode_mode","type":"umode_t","value":33261},{"name":"interpreter_pathname","type":"const char*","value":"/usr/lib64/ld-2.28.so"},{"name":"interpreter_dev","type":"dev_t","value":149},{"name":"interpreter_inode","type":"unsigned long","value":18975284},{"name":"interpreter_ctime","type":"unsigned long","value":1699952786674808884},{"name":"argv","type":"const char**","value":["/usr/bin/coreutils","--coreutils-prog-shebang=sleep","/usr/bin/sleep","999"]},{"name":"interp","type":"const char*","value":"/usr/bin/coreutils"},{"name":"stdin_type","type":"string","value":"S_IFCHR"},{"name":"stdin_path","type":"char*","value":"/dev/null"},{"name":"invoked_from_kernel","type":"int","value":0},{"name":"env","type":"const char**","value":null}]}


```

## Output of `tracee when executed`:

```
"level":"warn","ts":1699952901.596234,"msg":"KConfig: could not check enabled kconfig features","error":"could not read /boot/config-5.10.198-187.748.amzn2.x86_64: stat /boot/config-5.10.198-187.748.amzn2.x86_64: no such file or directory"}


```

### LINKED ISSUE #3824: Incorrect Container Tracking in TAS Cloud-Foundry Environment Due to Cgroup Mount Path Selection
## Description
There is an issue in the container tracking mechanism for existing containers within TAS environments. The current implementation selects the last matching record in `/proc/mounts` for cgroup mounts. However, this approach has encountered challenges, particularly in scenarios where the mount paths are listed in a reverse order compared to our initial assumptions.
As a result, existing containers are not being tracked at all in cloud-foundry environment.

**Observed Behavior:**

In specific setups, the order of mount paths in `/proc/mounts` differs from the usual pattern. The last record found, which our system currently relies on to identify the container's mount path, often does not correctly reflect the actual host mount. This leads to a failure in accurately detecting and tracking existing containers.

**Expected Behavior:**

The container tracking mechanism should reliably identify and track containers regardless of the order of mount paths in `/proc/mounts`. The correct host mount path should be determined in a way that is not dependent on the order of the records.

### Steps to Reproduce:

1. **Deploy an Application**: Start by running any application within a TAS (Tanzu Application Service) environment.

2. **Install Tracee**: Proceed to install Tracee in the same TAS environment.

3. **Observe Application Tracking**: After the installation of Tracee, observe the behavior related to application tracking. You may notice that the existing application, which was running prior to the Tracee installation, is not being tracked as expected.

**Suggested Fix:**

A potential solution is to modify the logic for selecting the mount path by considering the length of the path, rather than its position in the list. By selecting the longest mount path, we can achieve a more accurate representation of the host mounts.

## Output of `tracee version`:

```
v0.19.0
```

## Output of `uname -a`:

```
Linux 5c79c746-5998-4cb1-9a97-7bd3d46d76fe 6.2.0-37-generic #38~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Nov  2 18:01:13 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

```

### LINKED ISSUE #3826: Detect cgroup mounting through mountinfo instead of /proc/mounts
Our code to manage mounting of filesystems, in particular used for detecting the cgroup filesystem which tracee should track, is based on parsing the /proc/mounts file. This file only includes the target location of the mount in the internal file system in case we run in a container. By parsing the mountinfo file instead (in /proc/self/mountinfo), we can find out which source in the root filesystem was mounted. In these cases, we can see if tracee will accidentally choose a managed mount on an irrelevant filesystem (for example as is the case in #3689). This can give us a more accurate criteria for creating a managed mount.

### PR DESCRIPTION
**Pull request by NDStrahilevitz**
### 1. What the PR does

ae406df9c **optimize(enrich): skip non hid mkdir events**
9d97d239c **chore: improve logging in container related code**
96f0c9042 **fix: existing_container bad output**
04331580b **feat(mount): use mountinfo instead of /proc/mounts**
9d2206c4a **feat(cgroup): always switch to root cgroupns**

### 2. Explain how to test it

1. Deploy an EKS environment (v1.27+)
2. Deploy tracee with a rule for tracking container_create
3. Deploy a new pod
4. The two container_create events should be enriched (barring some race conditions)

### 3. Other comments

I've found that we have some race conditions right now where enrichment will finish but a container event will be sent before. This is another bug which should be tackled separately (either by fixing the condition itself, or changing how a container_create event is triggered).

Resolve #3826
Fix #3689
Fix #3824

### CODE REVIEW

File: pkg/cgroup/cgroup.go
Comment (geyslan) [inline pkg/cgroup/cgroup.go:89]: Is this thread returning to its previous cgroup ns somewhere? If not, wouldn't this logic be in a temp goroutine call? WDYT?

**Review by geyslan — COMMENTED** [review]
LGTM overall. I put some questions.

File: pkg/ebpf/events_enrich.go
Comment (NDStrahilevitz) [reply pkg/ebpf/events_enrich.go:108]: We've already asserted it in the prior if, but maybe just to ensure nothing weird is going on...

File: pkg/ebpf/events_enrich.go
Comment (NDStrahilevitz) [reply pkg/ebpf/events_enrich.go:115]: I would have to internally check `if err!=nil` anyway because I don't want to log in the other case. With that in mind, I think this is cleaner.

File: pkg/ebpf/events_enrich.go
Comment (NDStrahilevitz) [reply pkg/ebpf/events_enrich.go:116]: You're probably right, I can add this check in the PR.

File: pkg/ebpf/tracee.go
Comment (NDStrahilevitz) [reply pkg/ebpf/tracee.go:1644]: This is because of (this)[https://go.dev/blog/loopvar-preview]. We had other parts of the code affected by this before, there was a bug here (see the third commit). 
I don't find the syntax nicer personally, but maybe once this is moved from experimental we can go back to regular ranges.

File: pkg/events/usermode.go
Comment (NDStrahilevitz) [reply pkg/events/usermode.go:99]: Right, this is sort of a leftover from my debugging of the seen above bug. I've decided to keep it for readability.

File: pkg/mount/mount.go
Comment (NDStrahilevitz) [reply pkg/mount/mount.go:217]: Right, I need to change the function docs. With the new logic, any mount may suffice so long as it is actually on the host, so may as well pick the first one.

File: pkg/cgroup/cgroup.go
Comment (NDStrahilevitz) [reply pkg/cgroup/cgroup.go:85]: Force of habit... Thanks for catching.

File: pkg/cgroup/cgroup.go
Comment (NDStrahilevitz) [reply pkg/cgroup/cgroup.go:89]: No, supposedly golang can execute functions on different threads opportunistically. This ensures that the mounting process will occur in the same thread as the one we changed cgroupns in.
I've actually kept this code so that tracee is always deployed in the root cgroupns. Might be more appropriate to do in a cgo constructor when tracee executes, since that is done in a single threaded context and as such will propagate to tracee's threads as a whole.

Then again, removing this code hasn't affected functionality of the fix as far as i've tested. So maybe we should drop the commit. WDYT?

File: pkg/ebpf/events_enrich.go
Comment (geyslan) [reply pkg/ebpf/events_enrich.go:115]: We could check only err first to log. After doing the OR check.

File: pkg/cgroup/cgroup.go
Comment (geyslan) [reply pkg/cgroup/cgroup.go:89]: > Then again, removing this code hasn't affected functionality of the fix as far as i've tested. So maybe we should drop the commit. WDYT?

I wouldn't drop it, but I would insert a comment about the cgo constructor and/or that the current caller's goroutine will possibly be in a different cgroup mount ns than the batch. It would help us in the future.

**Review by geyslan — APPROVED** [review]
Approving for when ready.

**Comment by NDStrahilevitz** [comment]
@yanivagman Do you want to review this before I fix and merge?

**Review by yanivagman — APPROVED** [review]
LGTM after the suggested changes

File: pkg/mount/mount.go
Comment (geyslan) [reply pkg/mount/mount.go:226]: We're returning the first occurrence, right? But the function doc still mentions "the last".

File: pkg/cgroup/cgroup.go
Comment (NDStrahilevitz) [reply pkg/cgroup/cgroup.go]: Done

File: pkg/mount/mount.go
Comment (geyslan) [inline pkg/mount/mount.go:33]:
```suggestion
// from the host filesystem, and if not, it will mount it (in a temporary directory)
```

Comment (geyslan) [reply pkg/mount/mount.go:33]: After the tests and this fix, please merge it. Thanks!

**Pull request merged**
